src/lib/Microsoft.Health.Fhir.Ingest/Service/NormalizationEventConsumerService.cs

```diff
@@ -0,0 +1,218 @@
+            _templateManager = EnsureArg.IsNotNull(templateManager, nameof(templateManager));
+            _collector = EnsureArg.IsNotNull(collector, nameof(collector));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+            _retryPolicy = CreateRetryPolicy(logger);
+            _collectionTemplateFactory = EnsureArg.IsNotNull(collectionTemplateFactory, nameof(collectionTemplateFactory));
+            _exceptionTelemetryProcessor = EnsureArg.IsNotNull(exceptionTelemetryProcessor, nameof(exceptionTelemetryProcessor));
+
+            EventMetrics.SetConnectorOperation(ConnectorOperation.Normalization);
+        }
+
+        private (IContentTemplate template, DateTimeOffset timestamp) NormalizationTemplate { get; set; }
+
+        public async Task ConsumeAsync(IEnumerable<IEventMessage> events)
+        {
+            EnsureArg.IsNotNull(events);
+
+            await _retryPolicy.ExecuteAsync(async () => await ConsumeAsyncImpl(events));
+        }
+
+        private async Task<IContentTemplate> GetNormalizationTemplate()
+        {
+            await semaphore.WaitAsync();
+            try
+            {
+                if (NormalizationTemplate.template == null)
+                {
+                    _logger.LogTrace("Initializing normalization template from blob.");
+                    DateTimeOffset timestamp = DateTimeOffset.UtcNow;
+                    var content = await _templateManager.GetTemplateContentIfChangedSince(_templateDefinition);
+                    var templateContext = _collectionTemplateFactory.Create(content);
+                    templateContext.EnsureValid();
+                    NormalizationTemplate = (templateContext.Template, timestamp);
+                }
+                else
+                {
+                    DateTimeOffset updatedTimestamp = DateTimeOffset.UtcNow;
+                    var content = await _templateManager.GetTemplateContentIfChangedSince(_templateDefinition, NormalizationTemplate.timestamp);
+
+                    if (content != null)
+                    {
+                        _logger.LogTrace("New normalization template content detected, updating template.");
+                        var templateContext = _collectionTemplateFactory.Create(content);
+                        templateContext.EnsureValid();
+                        NormalizationTemplate = (templateContext.Template, updatedTimestamp);
+                    }
+                }
+
+                return NormalizationTemplate.template;
+            }
+            finally
+            {
+                semaphore.Release();
+            }
+        }
+
+        private async Task ConsumeAsyncImpl(IEnumerable<IEventMessage> events)
+        {
+            var template = await GetNormalizationTemplate();
+
+            var normalizationBatch = new List<(string sourcePartition, IMeasurement measurement)>(50);
```

Comment:
What is the mechanism that limits the number of `events` to 50?

Reply:
This doesn't limit the list to 50, this is just initial capacity of the list (the internal array size that backs the list).  If not set, the array size is 0 and then needs to be increased as elements are added.  If more than 50 elements, the capacity will be expanded.  

The intent is to set a reasonable starting capacity excessive new array allocations & array copies as the list grows.  I believe the implementation of List doubles the capacity when it exceeds the limit.

See https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.-ctor?view=net-6.0#system-collections-generic-list-1-ctor(system-int32) for more details.

Reply:
I was able to find the source code here https://source.dot.net/#System.Private.CoreLib/List.cs,421, when growing it does double the underlying array length.  If no capacity is set, when it first grows, the default array size is 4.